### ISSUE TITLE
CSS Selectors 4, :focus-visible.

### ISSUE THREAD

**Issue by robdodson** [w3ctag/design-reviews#233]
こんにちはTAG！

I'm requesting a TAG review of:

  - Name: CSS Selectors 4, :focus-visible pseudo class.
  - Specification URL: https://drafts.csswg.org/selectors-4/#the-focus-visible-pseudo
  - Explainer, Requirements Doc, or Example code: https://github.com/WICG/focus-visible/blob/master/explainer.md
  - Tests: Polyfill tests (https://github.com/WICG/focus-visible/tree/master/test), wpt will live in https://github.com/w3c/web-platform-tests/tree/master/css/selectors (not written yet).
  - Primary contacts: @robdodson, @alice

We'd prefer the TAG provide feedback as (please select one):

  - [ ] open issues in our Github repo for each point of feedback
  - [ ] open a single issue in our Github repo for the entire review
  - [x] leave review feedback as a comment in this issue and @-notify [github usernames]

**Comment by dbaron** [comment]
I think this is small and relatively non-controversial.  It's existed in Gecko for a while as `:-moz-focusring`.

I guess the biggest thing I wonder about when reading the specification is whether the conformance criteria should be as loose as they are, or whether they should be a little bit more specific.  I guess it's probably fine as they are.

So, basically, I think this is fine.

**Comment by bkardell** [comment]
I think that there are a few cases where we actually have verified through use of the polyfill and feedback from some fairly widely used sites/apps that we would kind of not want it to match exactly what :-moz-focusring does.  We don't want to get too over-specified about it but we were working on a pull that describes it better.  I think it might be worth looking at the tests, if the tests are up to date (honestly, @robdodson would know a lot better at this point)?  So... idk, fwiw, I do think we may want it to be just a little more wordy, at least in the note about intent if not actually specified strict conformance criteria?

**Comment by robdodson** [comment]
Yep the tests are up to date :) As Brian mentioned, we are working on a PR to better describe the intent and we might delve a bit more into conformance. We can follow up on this thread when we've refined that language.

**Comment by torgo** [comment]
session scheduled for Saturday on this

**Comment by travisleithead** [comment]
Maybe this is my lack of understanding, but if I target an element with both `:focus` *and* `:focus-visible` in a stylesheet, won't the `:focus` rule make the `:focus-visible` rule never work? (Since `:focus` always applies when an element is in focus, and `:focus-visible` conditionally applies?) So, if I, as a developer, want to start using `:focus-visible` in my stylesheet for UA's that support it, but keep `:focus` around for those that don't, I guess I must resort to @supports? Or is it understood somehow that `:focus-visible` wins over `:focus` assuming specificity is otherwise equal?

**Comment by travisleithead** [comment]
OK. Glad to get that confirmed. Thanks!

**Comment by robdodson** [comment]
Adding pseudo classes to @supports would be a huge help.

As Brian mentioned, without @supports, using [the polyfill](https://github.com/WICG/focus-visible) may be the easiest way to provide backwards compatibility. 

Another possible trick [suggested by the Paciello Group](https://developer.paciellogroup.com/blog/2018/03/focus-visible-and-backwards-compatibility/) would be to do something like:

```
button:focus { /* some exciting button focus styles */ }
button:focus:not(:focus-visible) {
    /* undo all the above focused button styles
       if the button has focus but the browser wouldn't normally
       show default focus styles */
}
button:focus-visible { /* some even *more* exciting button focus styles */ }
```
I haven't tested it across all browsers, but the author suggests that if the browser doesn't know about the pseudo class in the `:not()` bit, then it ignores the entire selector.

**Comment by torgo** [comment]
Still waiting on status / fate of w3c/csswg-drafts#2481 ... we will keep this open for now and come back to it.

**Comment by robdodson** [comment]
I chatted with @tabatkins today, he said he may want to take a stab at updating the spec text based on our discussion and some of the content in the linked PR.

**Comment by dbaron** [comment]
Seems that the CSS WG discussion has now moved to w3c/csswg-drafts#2897, which is still open.

**Comment by dbaron** [comment]
That PR is now merged, so the text is now [part of the spec](https://drafts.csswg.org/selectors-4/#the-focus-visible-pseudo).
